### ISSUE TITLE
 Use tf.function on train_step

### ISSUE THREAD

**Issue by axeldavy** [andreped/GradientAccumulator#102]
Hi there, this repo is very nice !

I've tried the Model wrapper on some tensorflow 2.6 code, and somehow it was using much more memory and being slower than without accumulation.
This got fixed by adding the @tf.function decorator on train_step.

However the function needs some modifications for it to work correctly I believe.
I removed the lines with "if self.first_call:" because I believe it will not give what you want with @tf.function.
There might be other changes that I'm not seeing. One should double check the updates are really exactly similar.

**Comment by andreped** [comment]
Hello, @axeldavy!

If you attempt to add the `tf.function` decorator to the `train_step` of the model wrapper, you should get an error/warning.

Basically, when suclassing the `keras.Model` you should **not** add decorators to the overloaded methods.

For instance, in [this](https://keras.io/guides/customizing_what_happens_in_fit/) tutorial by Francois Chollet (creator of Keras), decorators are not used.

As I believe you added decorators and it seemed to work fine in your setup, it would be nice to know stuff like: operating system used, python version, tensorflow version. The `bug_report` template should have had that covered.

**Comment by andreped** [comment]
I believe [this](https://github.com/tensorflow/tensorflow/issues/42119#issuecomment-1056109431) comment includes the error message you should observe, if you were to add the `tf.function` decorator. I believe you should have observed the same on single-GPU training.

**Comment by axeldavy** [comment]
I see, then it is due to the fact I'm using a custom training loop.
(https://www.tensorflow.org/guide/keras/writing_a_training_loop_from_scratch uses tf.function on a function they call train_step, but indeed it doesn't seem to be model.train_step, though it does a similar thing).

**Comment by andreped** [comment]
Are you overriding the `train_step` when making your custom training loop, or are you doing it externally?

**Comment by axeldavy** [comment]
I am overriding the `train_step` indeed. I have made a custom class inheriting my modified model already, and applying a few minor changes such that it would work properly.

The documentation made me believe using the Model wrapper would be easier than the optimizer, since the optimizer version has some compatibility limitations. I'll try that too and keep you updated.

**Comment by andreped** [comment]
> the optimizer version has some compatibility limitations

It is definitely not perfect, both solutions really, but I have gotten a lot faster in making them more seemless to use and more robust. However, the optimizer wrapper has not yet been as rigorously benchmarked, so I do not how well it works for various applications.

Importantly, which application are you working with? Note that I am currently experiencing some issues with NLP projects, as the Embedding layer, which depends on sparse updates, yields poorer approximation than other layers that use dense updates. However, for 95% of applications, it should work (jinx).

**Comment by andreped** [comment]
Also, see [this thread](https://github.com/andreped/GradientAccumulator/issues/23) regarding model subclassing. This is related to the `self.first_call` stuff you mentioned.

**Comment by axeldavy** [comment]
Ok, I see for the `self.first_call` stuff. Maybe this could have been done a different way by inserting the call somewhere else (compile() ?).
Also I read in some issues that the accumulated variables can appear in the list of saved/loaded weights. I believe self._self_setattr_tracking = False might help there.

**Comment by axeldavy** [comment]
Also regarding to your question, I'm working on object detection. I do not have issues with accuracy due to accumulating gradients to the best of my knowledge, but the automated loss scaling is important.

**Comment by andreped** [comment]
> Also I read in some issues that the accumulated variables can appear in the list of saved/loaded weights. I believe self._self_setattr_tracking = False might help there.

Thats interesting. This is likely related to [this](https://github.com/andreped/GradientAccumulator/issues/26#issuecomment-1536684312) issue. If you manage to get it working by disabling the tracking, please let me know! Would be of great interest. PR welcome! A unit test was made for exactly this issue (see [here](https://github.com/andreped/GradientAccumulator/blob/main/tests/test_param_count.py#L21)), if you wish to see if you can resolve the issue :]

> Also regarding to your question, I'm working on object detection. I do not have issues with accuracy due to accumulating gradients to the best of my knowledge, but the automated loss scaling is important.

Yes, then I believe it should work fine. The model wrapper should also work, as long as you handle subclassing correctly. Note that the model wrapper also supports mixed precision.

**Comment by andreped** [comment]
Note that if you are doing object detection, you likely want to use YOLOv8. See [this](https://keras.io/api/keras_cv/models/backbones/yolo_v8/) tutorial.
